### DISCUSSION TITLE
WARNING  No locally installed `turbo` found. Using version: 2.0.4.

### DISCUSSION THREAD

**Discussion by bestickley** [vercel/turborepo#8514] [Help]
### Summary

I have a PNPM monorepo where I have turbo installed in the root workspace and I then use turbo in all other workspaces. This worked fine for turbo v1, but now I'm getting the warning: " WARNING  No locally installed `turbo` found. Using version: 2.0.4.".

Does turbo repo v2 required/prefer you to install turbo in all child workspaces?

### Additional information

Relatedly, I'm getting the error:
```
pnpm turbo prune template-svc-core --docker --out-dir pruned/tmpl-svc-main-fn
 WARNING  No locally installed `turbo` found. Using version: 2.0.4.
  × could not resolve workspaces: missing packageManager field in package.json
  ╰─▶ missing packageManager field in package.json
```
but I have `packageManager` set in my root package.json, just not in a nested package.json. Do I have to set `packageManager` in all child workspaces?

**Comment by bestickley** [comment]
Strangely, this was cause by changing my pnpm-workspace.yaml from:

```yaml
packages:
  - "packages/**"
```
to
```yaml
packages:
  - "./packages/**"
```

Issue has been resolved. I've reverted the change to fix.

**Reply by anthonyshew** [reply]
Out of curiosity, what version of pnpm are you using?

**Reply by bestickley** [reply]
9.3.0

**Comment by timostamm** [comment]
Same with the "workspaces" field in package.json (for npm). E.g.

```json
{
  "name": "example",
  "private": true,
  "workspaces": [
    "./packages/foo",
    "./packages/bar",
  ]
}
```

Automatic package scoping in 2.0.4 does not support these paths. Running a task in `packages/foo` fails with the error message in the title.

Since the paths with leading `./` are supported in npm and yarn, I think that turbo should support them as well.

**Reply by anthonyshew** [reply]
Automatic Package Scoping is only syntatic sugar for `--filter`. Are you seeing `--filter`s misbehave as well?

Agree that `turbo` should support them as well. Could you open an Issue with a reproduction for this?

**Reply by timostamm** [reply]
See https://github.com/vercel/turbo/issues/8599

**Comment by sidarth-23** [comment]
I'm using pnpm version 9.5.0. I'm not sure why this error. Should I give packageManager for every application?

![image](https://github.com/vercel/turbo/assets/58748190/059a22c3-db41-44b5-b946-4f3dff340004)

![image](https://github.com/vercel/turbo/assets/58748190/927f19e2-6003-4f0b-bdb8-03e8db6a6345)

![image](https://github.com/vercel/turbo/assets/58748190/e708950f-2856-48cf-99ae-7205eff8e45b)

**Reply by sidarth-23** [reply]
If I give package manager to induvidual app, then this error pops up

![image](https://github.com/vercel/turbo/assets/58748190/f94475d1-0dc6-4b40-90c2-d11611989aef)

**Reply by anthonyshew** [reply]
It looks like the error might be misleading here. Turborepo doesn't support deeply nested packages: https://turbo.build/repo/docs/crafting-your-repository/structuring-a-repository#declaring-directories-for-packages

If you change your paths to `apps/*` and `packages/*` and add a `packageManager` field to your root `package.json` [like the examples](https://github.com/vercel/turbo/blob/4db451b33910e3982a70c3d16bf54d0ab317e5d5/examples/basic/package.json#L15), does this error get resolved?

**Reply by 0xNegative** [reply]
I'm having the same issue here. Currently attempting to use Docker.

* `packageManager` is in the root `package.json`.
* `pnpm-workspace.yaml` - I have tried both ways of adding the paths mentioned here (ex `./apps/**`, `apps/*`)

```bash
 => [pruner 5/7] COPY . .                                                                                         0.0s 
 => ERROR [pruner 6/7] RUN turbo prune --scope=web --docker                                                       0.4s 
------                                                                                                                 
 > [pruner 6/7] RUN turbo prune --scope=web --docker:                                                                  
0.343  WARNING  No locally installed `turbo` found. Using version: 2.0.6.                                              
0.348                                                                                                                  
0.348 Attention:                                                                                                       
0.348 Turborepo now collects completely anonymous telemetry regarding usage.
0.348 This information is used to shape the Turborepo roadmap and prioritize features.
0.348 You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
0.348 https://turbo.build/repo/docs/telemetry
0.348 
0.349   x could not resolve workspaces: missing packageManager field in package.json
0.349   `-> missing packageManager field in package.json
0.349 
------
Error: failed to fetch an image or build from source: error building: failed to solve: process "/bin/sh -c turbo prune --scope=$APP_PKG_NAME --docker" did not complete successfully: exit code: 1
```

Here's my Dockerfile for reference.

```dockerfile
ARG APP_DIRNAME=web
ARG APP_PKG_NAME=web

ARG NODE_VERSION=20.15.0
FROM node:${NODE_VERSION} as pruner
ARG APP_PKG_NAME
RUN apt-get update && apt-get install -y openssl
WORKDIR /app
RUN yarn global add turbo
COPY . .
RUN turbo prune --scope=$APP_PKG_NAME --docker
RUN find . -name "node_modules" -type d -prune -exec rm -rf '{}' +

FROM node:${NODE_VERSION} AS base
RUN apt-get update && apt-get install -y openssl
RUN corepack enable
WORKDIR /app
COPY .gitignore .gitignore
COPY --from=pruner /app/out/json/ .
COPY --from=pruner /app/out/pnpm-lock.yaml ./pnpm-lock.yaml
COPY --from=pruner /app/out/pnpm-workspace.yaml ./pnpm-workspace.yaml

FROM base AS dev-deps
WORKDIR /app
RUN pnpm install

FROM base AS production-deps
WORKDIR /app
ENV NODE_ENV production
RUN pnpm install --prod

FROM base AS builder
ARG APP_PKG_NAME
WORKDIR /app
RUN corepack enable
ENV NODE_ENV production
COPY --from=pruner /app/out/full/ .
COPY --from=dev-deps /app/ .
COPY --from=pruner /app/turbo.json turbo.json
RUN pnpm run build --filter=$APP_PKG_NAME...

FROM base AS runner
ARG APP_DIRNAME

ENV NODE_ENV production

COPY --from=pruner /app/out/full/ .
COPY --from=production-deps /app .
COPY --from=builder /app/apps/$APP_DIRNAME/build ./apps/$APP_DIRNAME/build
COPY --from=builder /app/apps/$APP_DIRNAME/public ./apps/$APP_DIRNAME/public

CMD ["pnpm", "run", "start"]
```

**Reply by ghost** [reply]
> It looks like the error might be misleading here. Turborepo doesn't support deeply nested packages: https://turbo.build/repo/docs/crafting-your-repository/structuring-a-repository#declaring-directories-for-packages
> 
> If you change your paths to `apps/*` and `packages/*` and add a `packageManager` field to your root `package.json` [like the examples](https://github.com/vercel/turbo/blob/4db451b33910e3982a70c3d16bf54d0ab317e5d5/examples/basic/package.json#L15), does this error get resolved? 

I already have packageManager in my root package.json. and i have tried even with the default workspace setting that turbo gives me. But still the same thing pops up

**Comment by 3dos** [comment]
Have the same-ish issue. I'm trying to upgrade from v1 to v2 and it works perfectly fine locally but for an unknown reason, it fails on CI when running `yarn build`.

```
x No package found with name 'xxx' in workspace
```

I'm using Yarn and here's a truncated version of my root `package.json`

```json
{
  "name": "xxx",
  "private": true,
  "scripts": {
    "build": "turbo build"
  },
  "devDependencies": {
    // ...
  },
  "engines": {
    "node": ">=20"
  },
  "packageManager": "yarn@4.1.1",
  "workspaces": [
    "apps/*",
    "packages/*"
  ],
  "config": {
    "commitizen": {
      "path": "cz-conventional-changelog"
    }
  }
}
```

I have a very basic `turbo.json`:

```json
{
  "$schema": "https://turbo.build/schema.json",
  "globalDependencies": ["**/.env.*local"],
  "tasks": {
    "build": {
      "outputs": ["dist/**"],
      "dependsOn": ["^build"]
    },
    "test": {
      "outputs": ["coverage/**"],
      "dependsOn": []
    },
    "lint": {},
    "dev": {
      "cache": false,
      "persistent": true
    },
    "clean": {
      "cache": false
    }
  }
}
```

**Reply by skahack** [reply]
@anthonyshew 
I don't know if this is the same situation, but I have reproduced the code and will share it with you.

https://github.com/skahack/repro-turborepo-discussions-8514

**Comment by anthonyshew** [comment]
@sidarth-2000 @3dos @0xNegative I am unable to reproduce this. If someone could post a full reproduction, that would be great.

**Reply by sidarth-23** [reply]
@anthonyshew I have moved on with other priority tasks. Will try to go through my pc if im able to find that

**Comment by Mingyang-Li** [comment]
@3dos Are you able to generate an `out` folder if you do `npx turbo prune --scope=YOUR_APP --docker` from terminal?

It works on my terminal, but not in docker. I'm still looking for a solution.

**Comment by pedroct92** [comment]
If you used the [Dockerfile](https://github.com/vercel/turborepo/tree/main/examples/with-docker/apps/web) form example repo. Please, make sure to run your build from the root of your repository. The command `COPY . .` is relative to whe the build is happening thus the error is saying not package manager it could be because it only copied the package.json for the app you are trying to build. 

The docs already mentions that but I wanted to pin that here as other people will soon face the same issue.

**Comment by samhh** [comment]
I see this warning when installing Turbo via Nix instead of npm/similar.

I was able to silence it by overriding the derivation in my shell, similar to [nixpkgs](https://github.com/NixOS/nixpkgs/blob/24.11/pkgs/by-name/tu/turbo/package.nix):

```nix
pkgs.symlinkJoin {
  name = "turbo-rewrapped";
  paths = [ pkgs.turbo-unwrapped ];
  nativeBuildInputs = [ pkgs.makeBinaryWrapper ];
  postBuild = ''
    wrapProgram "$out/bin/turbo" \
      --set TURBO_TELEMETRY_DISABLED 1 \
      --set TURBO_NO_UPDATE_NOTIFIER 1 \
      --set TURBO_GLOBAL_WARNING_DISABLED 1
    '';
}
```

**Comment by photomoose** [comment]
If anyone is coming here as a result of only having a global installation of Turbo, then if you also install Turbo locally in the repository, the warning goes away. This is documented here:

https://turborepo.com/docs/getting-started/installation#repository-installation

**Comment by earnubs** [comment]
In my case this error was caused when running Docker build from a path other than the monorepo root, in CDK. So something like this is wrong (and gives the packageManager error):
```js
ecs.ContainerImage.fromAsset(path.resolve('../../apps/foo'), { // the CDK code lives in a directory like `./infra/bar`
  file: 'Dockerfile',
  ...
}  
```

but this is correct (or, works):

```js
ecs.ContainerImage.fromAsset(path.resolve('../../'), {
  file: 'apps/foo/Dockerfile',
  ...
}  
```
